etl/src/main/java/org/pih/hivmigration/etl/sql/LabResultMigrator.groovy

```diff
@@ -0,0 +1,261 @@
+package org.pih.hivmigration.etl.sql
+
+class LabResultMigrator extends SqlMigrator {
+
+    @Override
+    def void migrate() {
+        executeMysql("Create staging table", '''
+            create table hivmigration_lab_results (
+              obs_id int PRIMARY KEY AUTO_INCREMENT,
+              source_patient_id int,
+              source_encounter_id int,
+              source_result_id int,
+              sample_id VARCHAR(20),
+              test_type VARCHAR(16), -- viral_load, CD4, tr, ppd, hematocrit
+              obs_datetime date,
+              value_numeric DOUBLE,
+              value_text VARCHAR(100),
+              value_boolean BOOLEAN,
+              vl_beyond_detectable_limit BOOLEAN,
+              vl_detectable_lower_limit DOUBLE,
+              uuid char(38),
+              KEY `source_encounter_id_idx` (`source_encounter_id`)
+            );
+        ''')
+
+        loadFromOracleToMySql('''
+            insert into hivmigration_lab_results
+               (source_patient_id,
+                source_encounter_id,
+                source_result_id,
+                sample_id,
+                test_type,
+                obs_datetime,
+                value_numeric,
+                value_text,
+                value_boolean,
+                vl_beyond_detectable_limit,
+                vl_detectable_lower_limit)
+            values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+        ''', '''
+            select   e.patient_id as source_patient_id,
+                     e.ENCOUNTER_ID as source_encounter_id, 
+                     r.RESULT_ID as source_result_id,
+                     r.sample_id, 
+                     lower(t.name) as test_type,
+                     to_char(e.encounter_date, 'yyyy-mm-dd') as obs_datetime,
+                     value as value_numeric, 
+                     value_string as value_text,   
+                     case when (value_p='t') then ( 1) 	
+                        else 0
+                        end as value_boolean,                
+                     case when (r.RESULT_UNDETECTABLE='t') then ( 1) 	
+                        else 0
+                        end as vl_beyond_detectable_limit,                  
+                     r.RESULT_UNDETECTABLE_RANGE as vl_detectable_lower_limit
+            from     hiv_demographics_real d, hiv_encounters e, hiv_lab_results r, hiv_lab_tests t
+            where    d.patient_id = e.patient_id
+            and      e.encounter_id = r.encounter_id
+            and      r.lab_test_id = t.lab_test_id  
+            order by d.patient_id desc, e.encounter_id desc
+        ''')
+
+        executeMysql("Add UUIDs", "UPDATE hivmigration_lab_results SET uuid = uuid();")
+
+        executeMysql("Load viral load results as observations", '''
+            --
+            -- Create construct obs
+            --
+            -- These have to be created first, since their obs_ids are fixed. The obs group elements for
```

Comment:
When you say obs ids are fixed, you mean that they need to link back to the keys created in the hivmigration_labs_result, correct?

Reply:
Yes